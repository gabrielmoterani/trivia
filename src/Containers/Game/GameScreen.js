import React from 'react';
import { connect } from 'react-redux';
import { Layout, Text } from 'react-native-ui-kitten';
import BaseLayout from '../../Components/BaseLayout';
import AnswerChooser from '../../Components/AnswerChooser';
import GameActions from '../../Stores/Game/Actions';
import GameSelectors from '../../Stores/Game/Selectors';
import Style from './GameScreenStyle';

export function GameScreen(props) {
  return (
    <BaseLayout>
      <Layout style={Style.question}>
        <Text category="c1">{props.currentQuestionNumber} of {props.questionsSize}</Text>
        <Text category="c2">{props.currentQuestion.category}</Text>
        <Text category="h4">{props.currentQuestion.question}</Text>
      </Layout>
      <AnswerChooser
        currentQuestion={props.currentQuestion}
        currentAnswer={props.currentAnswer}
        onChoose={props.chooseAnswer}
      />
    </BaseLayout>
  );
}

const mapStateToProps = (state) => ({
  questionsSize: GameSelectors.questionsSize(state),
  currentQuestionNumber: GameSelectors.currentQuestionNumber(state),
  currentQuestion: GameSelectors.currentQuestion(state),
  currentAnswer: GameSelectors.currentAnswer(state),
});

const mapDispatchToProps = (dispatch) => ({
  chooseAnswer: (answer) => dispatch(GameActions.chooseGameCurrentAnswer(answer)),
});

export default connect(mapStateToProps, mapDispatchToProps)(GameScreen);
