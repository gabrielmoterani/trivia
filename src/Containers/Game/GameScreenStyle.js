import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  question: {
    paddingHorizontal: 25,
  },
  answer: {
    flexDirection: 'row',
    marginTop: 50,
    paddingHorizontal: 25,
    alignSelf: 'center',
  },
});
