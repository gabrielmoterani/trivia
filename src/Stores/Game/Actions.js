import { createActions } from 'reduxsauce';

const { Types, Creators } = createActions({
  newGame: null,
  updateGameQuestions: ['questions'],
  setGameCurrentQuestion: ['index', 'question'],
  chooseGameCurrentAnswer: ['answer'],
  setGameCurrentAnswer: ['answer', 'correct'],
  nextGameQuestion: null,
  resetGame: null,
});

export const GameTypes = Types;
export default Creators;
