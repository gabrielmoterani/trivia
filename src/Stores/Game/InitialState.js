import { Map, List } from 'immutable';

export const INITIAL_STATE = Map({
  loading: false,
  questions: List(),
  currentQuestion: Map(),
  currentQuestionIndex: null,
  answers: List(),
  currentAnswer: Map(),
});
