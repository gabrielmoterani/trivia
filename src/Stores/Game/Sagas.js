import { put, select, delay } from 'redux-saga/effects';
import { Alert } from 'react-native';
import i18n from '../../Locales/i18n';
import TriviaApiService from '../../Services/TriviaApiService';
import NavigationService from '../../Services/NavigationService';
import GameActions from './Actions';

export function* newGame() {
  try {
    // fetch questions from OTDB
    const questions = yield TriviaApiService.fetchQuestions();
    yield put(GameActions.updateGameQuestions(questions));

    // init first question
    yield put(GameActions.nextGameQuestion());

    // begin game!
    NavigationService.navigateAndReset('GameScreen');
  } catch (e) {
    // if fetching from OTDB fails, alert user about error and reset game
    const errorMsg = e && e.message || i18n.t('general.unknown');
    const alertMsg = `${i18n.t('general.errorMessage')}\n(${errorMsg})`;
    Alert.alert('Oooops :(', alertMsg, [], { cancelable: true });

    yield put(GameActions.resetGame());
  }
}

export function* chooseAnswer({ answer }) {
  // check if question is already answered
  const getCurrentAnswer = (state) => state.game.get('currentAnswer');
  const currentAnswer = yield select(getCurrentAnswer);
  if (currentAnswer.size) {
    return; // if already answered, ignore command;
  }

  // get current question to check if answer is correct
  const getCurrentQuestion = (state) => state.game.get('currentQuestion');
  const currentQuestion = yield select(getCurrentQuestion);

  // check if answer matches correct answer
  const isCorrect = answer === currentQuestion.get('correct_answer');

  // set current answer
  yield put(GameActions.setGameCurrentAnswer(answer, isCorrect));

  // wait a bit and move on to next question
  yield delay(2000);
  yield put(GameActions.nextGameQuestion());
}

export function* nextGameQuestion() {
  // get current question index
  const getCurrentQuestionIndex = (state) => state.game.get('currentQuestionIndex');
  const currentQuestionIndex = yield select(getCurrentQuestionIndex);
  const nextQuestionIndex = currentQuestionIndex === null ? 0 : currentQuestionIndex + 1;

  // get questions
  const getQuestions = (state) => state.game.get('questions');
  const questions = yield select(getQuestions);

  // if there is no more questions, "game over!" and go to results screen
  if (nextQuestionIndex >= questions.size) {
    return NavigationService.navigateAndReset('ResultScreen');
  }

  const currentQuestion = questions.get(nextQuestionIndex);
  yield put(GameActions.setGameCurrentQuestion(nextQuestionIndex, currentQuestion));
}

export function resetGame() {
  // reset game state and go to main screen
  NavigationService.navigateAndReset('MainScreen');
}
