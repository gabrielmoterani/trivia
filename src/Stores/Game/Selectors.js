import { createSelector } from 'reselect';

const questionsSelector = state => state.game.get('questions');
const answersSelector = state => state.game.get('answers');
const currentQuestionIndexSelector = state => state.game.get('currentQuestionIndex');
const currentQuestionSelector = state => state.game.get('currentQuestion');
const currentAnswerSelector = state => state.game.get('currentAnswer');

const questionsSize = createSelector(
  questionsSelector,
  (questions) => questions.size
);

const currentQuestion = createSelector(
  currentQuestionSelector,
  (question) => question.toJS()
);

const currentQuestionNumber = createSelector(
  currentQuestionIndexSelector,
  (questionIndex) => questionIndex ? questionIndex + 1 : 1
);

const currentAnswer = createSelector(
  currentAnswerSelector,
  (answer) => answer.toJS()
);

const nextQuestionIndex = createSelector(
  currentQuestionIndexSelector,
  questionsSize,
  (currentQuestionIndex, limit) => {
    const nextIndex = currentQuestionIndex === null ? 0 : currentQuestionIndex + 1;

    if (nextIndex >= limit) {
      return false;
    }

    return nextIndex;
  }
);

const nextQuestion = createSelector(
  nextQuestionIndex,
  questionsSelector,
  (questionIndex, questions) => {
    if (questionIndex === false) {
      return false;
    }

    return questions.get(questionIndex);
  }
);

const allAnswers = createSelector(
  answersSelector,
  (answers) => answers.toJS()
);

const score = createSelector(
  answersSelector,
  (answers) => answers.count(answer => answer.get('correct'))
);

export default {
  questionsSize,
  currentQuestion,
  currentQuestionNumber,
  currentAnswer,
  nextQuestionIndex,
  nextQuestion,
  allAnswers,
  score,
};
