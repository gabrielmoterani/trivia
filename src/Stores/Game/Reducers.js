import { createReducer } from 'reduxsauce';
import { Map, List } from 'immutable';
import { INITIAL_STATE } from './InitialState';
import { GameTypes } from './Actions';

export const newGame = (state) => {
  return state.merge({
    loading: true,
    currentScore: 0,
  });
};

export const updateGameQuestions = (state, { questions }) => {
  return state.merge({
    questions: List(questions),
  });
};

export const setGameCurrentQuestion = (state, { index, question }) => {
  return state.merge({
    loading: false,
    currentQuestion: Map(question),
    currentQuestionIndex: index,
    currentAnswer: Map(),
  });
};

export const setGameCurrentAnswer = (state, { answer, correct }) => {
  const currentScore = state.get('currentScore');
  const currentAnswers = state.get('answers');
  const currentAnswer = Map({
    answer,
    correct,
    question: state.get('currentQuestion'),
    questionIndex: state.get('currentQuestionIndex'),
  });

  return state.merge({
    loading: false,
    currentAnswer: Map({ answer, correct }),
    currentScore: correct ? currentScore + 1 : currentScore, // if it's correct, increase score
    answers: currentAnswers.push(currentAnswer),
  });
};

export const resetGame = () => INITIAL_STATE;

export const reducer = createReducer(INITIAL_STATE, {
  [GameTypes.NEW_GAME]: newGame,
  [GameTypes.UPDATE_GAME_QUESTIONS]: updateGameQuestions,
  [GameTypes.SET_GAME_CURRENT_QUESTION]: setGameCurrentQuestion,
  [GameTypes.SET_GAME_CURRENT_ANSWER]: setGameCurrentAnswer,
  [GameTypes.RESET_GAME]: resetGame,
});
