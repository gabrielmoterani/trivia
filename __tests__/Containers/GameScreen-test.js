import React from 'react';
import { shallow } from 'enzyme';
import { GameScreen } from '../../src/Containers/Game/GameScreen';

const initialProps = {
  questionsSize: 1,
  currentQuestionNumber: 1,
  currentQuestion: {
    category: 'Test Category',
    question: 'Is this a test?',
  },
  currentAnswer: {},
  chooseAnswer: jest.fn(),
};

const setup = (props = initialProps) => shallow(<GameScreen {...props} />);

describe('Test GameScreen container', () => {
  const wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have answer chooser', () => {
    const answers = wrapper.find('AnswerChooser');
    expect(answers).toExist();
  });
});
