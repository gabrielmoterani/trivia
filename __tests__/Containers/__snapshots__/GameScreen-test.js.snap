// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GameScreen container should render properly 1`] = `
<BaseLayout>
  <LayoutComponent
    style={
      Object {
        "paddingHorizontal": 25,
      }
    }
  >
    <TextComponent
      category="c1"
    >
      1
       of 
      1
    </TextComponent>
    <TextComponent
      category="c2"
    >
      Test Category
    </TextComponent>
    <TextComponent
      category="h4"
    >
      Is this a test?
    </TextComponent>
  </LayoutComponent>
  <AnswerChooser
    currentAnswer={Object {}}
    currentQuestion={
      Object {
        "category": "Test Category",
        "question": "Is this a test?",
      }
    }
    onChoose={[MockFunction]}
  />
</BaseLayout>
`;
